src/Common/tests/System/Net/Http/LoopbackServer.cs

```diff
@@ -85,193 +78,138 @@ public static Task CreateServerAndClientAsync(Func<Uri, Task> clientFunc, Func<S
-                await writer.WriteAsync(serverResponse).ConfigureAwait(false);
-                return true;
-            }
-
-            return false;
+            return AcceptConnectionSendResponseAndCloseAsync(GetDefaultHttpResponse());
         }
 
-        private static string ComputeWebSocketHandshakeSecurityAcceptValue(string secWebSocketKey)
+        public async Task AcceptConnectionAsync(Func<Connection, Task> funcAsync)
         {
-            // GUID specified by RFC 6455.
-            const string Rfc6455Guid = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
-            string combinedKey = secWebSocketKey + Rfc6455Guid;
-
-            // Use of SHA1 hash is required by RFC 6455.
-            SHA1 sha1Provider = new SHA1CryptoServiceProvider();
-            byte[] sha1Hash = sha1Provider.ComputeHash(Encoding.UTF8.GetBytes(combinedKey));
-            return Convert.ToBase64String(sha1Hash);
-        }
+            Socket s = await _listenSocket.AcceptAsync().ConfigureAwait(false);
 
-        public static async Task<List<string>> AcceptSocketAsync(Socket server, Func<Socket, Stream, StreamReader, StreamWriter, Task<List<string>>> funcAsync, Options options = null)
-        {
-            options = options ?? new Options();
-            Socket s = await server.AcceptAsync().ConfigureAwait(false);
-            s.NoDelay = true;
-            try
+            using (s)
             {
+                s.NoDelay = true;
+
                 Stream stream = new NetworkStream(s, ownsSocket: false);
-                if (options.UseSsl)
+                if (_options.UseSsl)
                 {
                     var sslStream = new SslStream(stream, false, delegate { return true; });
                     using (var cert = Configuration.Certificates.GetServerCertificate())
                     {
                         await sslStream.AuthenticateAsServerAsync(
                             cert,
                             clientCertificateRequired: true, // allowed but not required
-                            enabledSslProtocols: options.SslProtocols,
+                            enabledSslProtocols: _options.SslProtocols,
                             checkCertificateRevocation: false).ConfigureAwait(false);
                     }
                     stream = sslStream;
                 }
 
-                using (var reader = new StreamReader(stream, Encoding.ASCII))
-                using (var writer = new StreamWriter(options?.ResponseStreamWrapper?.Invoke(stream) ?? stream, Encoding.ASCII) { AutoFlush = true })
+                if (_options.StreamWrapper != null)
                 {
-                    return await funcAsync(s, stream, reader, writer).ConfigureAwait(false);
+                    stream = _options.StreamWrapper(stream);
                 }
-            }
-            finally
-            {
-                try
-                {
-                    s.Shutdown(SocketShutdown.Send);
```

Comment:
Because it shouldn't be necessary.  To the extent that it seems to help with test reliability, it appears to mask actual test issues by causing timing delays.  This doesn't seem like a good thing.

There was at least one test case where problems cropped up because the server was not sending Content-Length: 0, even though there was no response body.  Fixing the test made this problem go away.

Reply:
> Because it shouldn't be necessary

Without it, the client isn't guaranteed to receive all sent data, no?

Reply:
> Without it, the client isn't guaranteed to receive all sent data, no?

Calling Close on the socket does not cause the buffered data to be discarded.  The socket will continue to send and retry any buffered data, up to the linger timeout, in the background

See https://msdn.microsoft.com/en-us/library/windows/desktop/ms738547(v=vs.85).aspx for example, though this isn't the most lucid writeup.

Shutdown is really only useful if you want to do a half-close, i.e. stop sending data but continue to read it.  In the HTTP server case, when the server decides to disconnect, it's not interested in any more data from the client, so shutdown is not useful.

At least, that's my understanding.

Reply:
(That said, shutdown may have local-only effects that are useful, like causing a pending operation to cancel.  My point is that in these scenarios, it shouldn't affect what happens on the wire.)

Reply:
You may very well be right, but there are lots of comments on the web about how this is observably not the case (maybe just due to OS bugs). And from what we've observed, especially in CI, it's also suspect. I suggest don't remove it as part of this refactoring change, as I would not be at all surprised if we saw a resulting increase in spurious test failures. At least submit it as its own dedicated change, maybe a bit after this change guess through and is shown stable.

> Linger

Maybe it's the interaction of these that was an issue, with linger timeout set to 0 in some places IIRC.

Reply:
Thanks :)